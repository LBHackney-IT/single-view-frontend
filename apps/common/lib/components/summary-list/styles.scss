@import "~lbh-frontend/lbh/base";
@import "~lbh-frontend/lbh/components/lbh-summary-list/summary-list";

.govuk-summary-list.mtfh-summary-list--inline {
  font-size: 100%;
  width: auto;
  display: flex;
  flex-direction: column;
  margin-top: 1.5rem;

  .govuk-summary-list__row {
    display: flex;
    margin: 0;
  }

  .govuk-summary-list__key,
  .govuk-summary-list__value,
  .govuk-summary-list__actions {
    width: auto;
  }

  .govuk-summary-list__key,
  .govuk-summary-list__value,
  .govuk-summary-list__actions {
    padding: 0;
  }

  .govuk-summary-list__value,
  .govuk-summary-list__actions {
    margin-left: 0.5rem;
  }
}
