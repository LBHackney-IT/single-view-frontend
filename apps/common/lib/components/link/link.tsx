import React, { forwardRef } from "react";

import cn from "classnames";

import { widthOverrides } from "../../utils";

import type * as Polymorphic from "@radix-ui/react-polymorphic";
import "./styles.scss";

export interface LinkProps {
  variant?: "link" | "danger" | "text-colour" | "muted" | "back-link" | "native";
  isExternal?: boolean;
  override?: number;
}

export type LinkComponent = Polymorphic.ForwardRefComponent<"a", LinkProps>;

export const Link: LinkComponent = forwardRef(function Link(
  {
    as: LinkComp = "a",
    variant = "link",
    isExternal = false,
    className,
    rel,
    target,
    override,
    ...props
  },
  ref,
) {
  const linkClasses = cn(
    variant !== "native" && {
      "govuk-link lbh-link": variant !== "back-link",
      "govuk-back-link lbh-back-link": variant === "back-link",
      [`lbh-link--${variant}`]: variant !== "link" && variant !== "back-link",
      "lbh-link--no-visited-state": !isExternal,
    },
    widthOverrides(override),
    className,
  );
  return (
    // eslint-disable-next-line react/jsx-no-target-blank
    <LinkComp
      ref={ref}
      className={linkClasses}
      rel={isExternal ? "noopener noreferrer" : rel}
      target={isExternal ? "_blank" : target}
      {...props}
    />
  );
});
