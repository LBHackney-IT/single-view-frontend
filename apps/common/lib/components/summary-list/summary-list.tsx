import React, {
  Children,
  ComponentPropsWithoutRef,
  ReactElement,
  cloneElement,
  forwardRef,
  isValidElement,
  useMemo,
} from "react";

import cn from "classnames";

import { widthOverrides } from "../../utils";
import "./styles.scss";

export interface SummaryListItemProps extends ComponentPropsWithoutRef<"div"> {
  title: string;
  children?: ReactElement | string | string[] | null;
  actions?: ReactElement | ReactElement[];
  overrides?: number[];
  fallback?: string;
}

export const SummaryListItem = forwardRef<HTMLDivElement, SummaryListItemProps>(
  function SummaryListItem(
    { title, actions, children, className, fallback, overrides = [], ...props },
    ref,
  ) {
    const value = useMemo(
      () => (typeof children === "string" ? children.trim() : children),
      [children],
    );

    return (
      <div ref={ref} className={cn("govuk-summary-list__row", className)} {...props}>
        <dt className={cn("govuk-summary-list__key", widthOverrides(overrides[0]))}>
          {title}
        </dt>
        <dd className={cn("govuk-summary-list__value", widthOverrides(overrides[1]))}>
          {value || fallback || "N/A"}
        </dd>
        {actions && (
          <dd className={cn("govuk-summary-list__actions", widthOverrides(overrides[2]))}>
            <ul className="govuk-summary-list__actions-list">
              {Children.map(actions, (action) => (
                <li key={action.key} className="govuk-summary-list__actions-list-item">
                  {action}
                </li>
              ))}
            </ul>
          </dd>
        )}
      </div>
    );
  },
);

type SummaryListChild =
  | ReactElement<SummaryListItemProps>
  | ReactElement<SummaryListItemProps>[]
  | null;

export interface SummaryListProps extends ComponentPropsWithoutRef<"dl"> {
  variant?: "base" | "border" | "inline";
  overrides?: number[];
  children: SummaryListChild | SummaryListChild[];
}

export const SummaryList = forwardRef<HTMLDListElement, SummaryListProps>(
  function SummaryList(
    { variant = "base", className, overrides, children, ...props },
    ref,
  ) {
    return (
      <dl
        ref={ref}
        className={cn(
          "govuk-summary-list",
          { "govuk-summary-list--no-border": variant !== "border" },
          { "mtfh-summary-list--inline": variant === "inline" },
          "lbh-summary-list",
          className,
        )}
        {...props}
      >
        {Children.map(
          children,
          (child, index) =>
            child &&
            isValidElement<SummaryListItemProps>(child) &&
            cloneElement(child, {
              overrides:
                !child.props.overrides && index === 0 ? overrides : child.props.overrides,
            }),
        )}
      </dl>
    );
  },
);
