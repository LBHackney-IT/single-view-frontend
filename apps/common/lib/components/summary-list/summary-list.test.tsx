import React from "react";

import { render, testA11y } from "@hackney/mtfh-test-utils";

import { Link } from "../link";
import { SummaryList, SummaryListItem } from "./summary-list";

test("it renders correctly", async () => {
  const { container } = render(
    <SummaryList>
      <SummaryListItem title="Person">Test</SummaryListItem>
    </SummaryList>,
  );

  expect(container).toMatchSnapshot();
  await testA11y(container);
});

test("it renders correctly with fallback", () => {
  const { container } = render(
    <SummaryList>
      <SummaryListItem title="Email" />
      <SummaryListItem title="Mobile"> </SummaryListItem>
      <SummaryListItem title="Phone">{null}</SummaryListItem>
      <SummaryListItem title="Address">{undefined}</SummaryListItem>
    </SummaryList>,
  );

  expect(container).toMatchSnapshot();
});

test("it renders correctly with single action", async () => {
  const { container } = render(
    <SummaryList>
      <SummaryListItem title="First Name" actions={<Link href="/">Edit</Link>}>
        John
      </SummaryListItem>
      <SummaryListItem title="Last Name">Smith</SummaryListItem>
    </SummaryList>,
  );

  expect(container).toMatchSnapshot();
  await testA11y(container);
});

test("it renders correctly with multiple action", async () => {
  const { container } = render(
    <SummaryList>
      <SummaryListItem
        title="First Name"
        actions={[
          <Link key="edit" href="/edit">
            Edit
          </Link>,
          <Link key="remove" href="/remove">
            Remove
          </Link>,
        ]}
      >
        John
      </SummaryListItem>
      <SummaryListItem title="Last Name">Smith</SummaryListItem>
    </SummaryList>,
  );

  expect(container).toMatchSnapshot();
  await testA11y(container);
});

test("it renders correctly with columns", () => {
  const { container } = render(
    <SummaryList overrides={[1 / 2, 1 / 4, 1 / 2]}>
      <SummaryListItem
        title="First Name"
        actions={[
          <Link key="edit" href="/edit">
            Edit
          </Link>,
          <Link key="remove" href="/remove">
            Remove
          </Link>,
        ]}
      >
        John
      </SummaryListItem>
      <SummaryListItem title="Last Name">Smith</SummaryListItem>
      <SummaryListItem title="Gender">M</SummaryListItem>
    </SummaryList>,
  );

  expect(container).toMatchSnapshot();
});
