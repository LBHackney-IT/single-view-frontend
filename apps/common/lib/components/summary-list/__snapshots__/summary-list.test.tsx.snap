// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders correctly 1`] = `
<div>
  <dl
    class="govuk-summary-list govuk-summary-list--no-border lbh-summary-list"
  >
    <div
      class="govuk-summary-list__row"
    >
      <dt
        class="govuk-summary-list__key"
      >
        Person
      </dt>
      <dd
        class="govuk-summary-list__value"
      >
        Test
      </dd>
    </div>
  </dl>
</div>
`;

exports[`it renders correctly with columns 1`] = `
<div>
  <dl
    class="govuk-summary-list govuk-summary-list--no-border lbh-summary-list"
  >
    <div
      class="govuk-summary-list__row"
    >
      <dt
        class="govuk-summary-list__key govuk-!-width-one-half"
      >
        First Name
      </dt>
      <dd
        class="govuk-summary-list__value govuk-!-width-one-quarter"
      >
        John
      </dd>
      <dd
        class="govuk-summary-list__actions govuk-!-width-one-half"
      >
        <ul
          class="govuk-summary-list__actions-list"
        >
          <li
            class="govuk-summary-list__actions-list-item"
          >
            <a
              class="govuk-link lbh-link lbh-link--no-visited-state"
              href="/edit"
            >
              Edit
            </a>
          </li>
          <li
            class="govuk-summary-list__actions-list-item"
          >
            <a
              class="govuk-link lbh-link lbh-link--no-visited-state"
              href="/remove"
            >
              Remove
            </a>
          </li>
        </ul>
      </dd>
    </div>
    <div
      class="govuk-summary-list__row"
    >
      <dt
        class="govuk-summary-list__key"
      >
        Last Name
      </dt>
      <dd
        class="govuk-summary-list__value"
      >
        Smith
      </dd>
    </div>
    <div
      class="govuk-summary-list__row"
    >
      <dt
        class="govuk-summary-list__key"
      >
        Gender
      </dt>
      <dd
        class="govuk-summary-list__value"
      >
        M
      </dd>
    </div>
  </dl>
</div>
`;

exports[`it renders correctly with fallback 1`] = `
<div>
  <dl
    class="govuk-summary-list govuk-summary-list--no-border lbh-summary-list"
  >
    <div
      class="govuk-summary-list__row"
    >
      <dt
        class="govuk-summary-list__key"
      >
        Email
      </dt>
      <dd
        class="govuk-summary-list__value"
      >
        N/A
      </dd>
    </div>
    <div
      class="govuk-summary-list__row"
    >
      <dt
        class="govuk-summary-list__key"
      >
        Mobile
      </dt>
      <dd
        class="govuk-summary-list__value"
      >
        N/A
      </dd>
    </div>
    <div
      class="govuk-summary-list__row"
    >
      <dt
        class="govuk-summary-list__key"
      >
        Phone
      </dt>
      <dd
        class="govuk-summary-list__value"
      >
        N/A
      </dd>
    </div>
    <div
      class="govuk-summary-list__row"
    >
      <dt
        class="govuk-summary-list__key"
      >
        Address
      </dt>
      <dd
        class="govuk-summary-list__value"
      >
        N/A
      </dd>
    </div>
  </dl>
</div>
`;

exports[`it renders correctly with multiple action 1`] = `
<div>
  <dl
    class="govuk-summary-list govuk-summary-list--no-border lbh-summary-list"
  >
    <div
      class="govuk-summary-list__row"
    >
      <dt
        class="govuk-summary-list__key"
      >
        First Name
      </dt>
      <dd
        class="govuk-summary-list__value"
      >
        John
      </dd>
      <dd
        class="govuk-summary-list__actions"
      >
        <ul
          class="govuk-summary-list__actions-list"
        >
          <li
            class="govuk-summary-list__actions-list-item"
          >
            <a
              class="govuk-link lbh-link lbh-link--no-visited-state"
              href="/edit"
            >
              Edit
            </a>
          </li>
          <li
            class="govuk-summary-list__actions-list-item"
          >
            <a
              class="govuk-link lbh-link lbh-link--no-visited-state"
              href="/remove"
            >
              Remove
            </a>
          </li>
        </ul>
      </dd>
    </div>
    <div
      class="govuk-summary-list__row"
    >
      <dt
        class="govuk-summary-list__key"
      >
        Last Name
      </dt>
      <dd
        class="govuk-summary-list__value"
      >
        Smith
      </dd>
    </div>
  </dl>
</div>
`;

exports[`it renders correctly with single action 1`] = `
<div>
  <dl
    class="govuk-summary-list govuk-summary-list--no-border lbh-summary-list"
  >
    <div
      class="govuk-summary-list__row"
    >
      <dt
        class="govuk-summary-list__key"
      >
        First Name
      </dt>
      <dd
        class="govuk-summary-list__value"
      >
        John
      </dd>
      <dd
        class="govuk-summary-list__actions"
      >
        <ul
          class="govuk-summary-list__actions-list"
        >
          <li
            class="govuk-summary-list__actions-list-item"
          >
            <a
              class="govuk-link lbh-link lbh-link--no-visited-state"
              href="/"
            >
              Edit
            </a>
          </li>
        </ul>
      </dd>
    </div>
    <div
      class="govuk-summary-list__row"
    >
      <dt
        class="govuk-summary-list__key"
      >
        Last Name
      </dt>
      <dd
        class="govuk-summary-list__value"
      >
        Smith
      </dd>
    </div>
  </dl>
</div>
`;
